### ISSUE TITLE
AuthCodeGrant validateAuthorizationRequest uses defaultClientRedirectUri

### ISSUE THREAD

**Issue by hafkenscheid** [thephpleague/oauth2-server#1125]
Case: 
- A client with 
   - multiple redirect URIs: ``http://localhost1 http://localhost2 http://localhost3``
   - none, one or many scopes: ``scope1 scope2``
- A request with:
   - a valid redirect URI (not being the first one): ``http://localhost2``
   - an invalid scope ``scope3``

```
<?php
//...

namespace League\OAuth2\Server\Grant;

//...

class AuthCodeGrant extends AbstractAuthorizeGrant
{

//...

    public function validateAuthorizationRequest(ServerRequestInterface $request)
    {

//...

        $defaultClientRedirectUri = \is_array($client->getRedirectUri())
            ? $client->getRedirectUri()[0]
            : $client->getRedirectUri();

        $scopes = $this->validateScopes(
            $this->getQueryStringParameter('scope', $request, $this->defaultScope),
            $defaultClientRedirectUri
        );

//...
```
Result: 
- A response to the defaultClientUri, being in this case ``http://localhost1``

So, the response is going to another URL, despite the fact that the redirect_uri is valid. 
Is this a bug or a feature?

**Comment by Sephster** [comment]
If you have multiple redirect URIs against a client, you must specify which one you want the request to be redirected to. If this is not specified in the request as the `redirect_uri` parameter, we just take the first one in the array and redirect to that.

It might be more appropriate to enforce clients to specify the redirect uri they want us to use if they've registered multiple ones but this would be a breaking change.

Can I just check - did you explicitly provide the redirect_uri parameter in your request?

**Comment by hafkenscheid** [comment]
I do explicitly set the redirect_uri parameter (``http://localhost2``), it is validated, and valid. 

When this function starts validating the scopes, it ignores the validated redirect_uri and starts using the default. That goes well if the scopes are valid, but in the case of invalid scopes, the error only holds the default. And the error handler than returns to the default instead of the explicitly set and valid redirect_uri.

**Comment by Sephster** [comment]
I see. Yep, this looks to be a bug. Previously the library forced clients to specify a redirect_uri on every call. The default URI was added so that clients didn't have to specify the redirect URI. Looks like we need to modify this so it only uses the default if a valid redirect URI hasn't been provided. Thanks for pointing this out

**Comment by Sephster** [comment]
Fixed in PR #1126. I will aim to create a bug fix release tomorrow. Cheers again for reporting.
